Test Explorer Support via VSTest

farlee2121 opened the pull request:
# Test Explorer VSTest Migration

NOTE: It looks like a lot of changed files, but many of them are sample test projects or .fsi files

## Motivation

- [Microsoft is releasing a new test platform, and it doesn't support the `dotnet test` parameter we use](https://github.com/ionide/ionide-vscode-fsharp/issues/2069)
- [Access to standardized test code locations](https://github.com/ionide/ionide-vscode-fsharp/pull/2000)
- Ability to write automated tests against test explorer code
- Ability to share test explorer features across FSAC clients
- [Improve test discovery speed](https://github.com/ionide/ionide-vscode-fsharp/issues/2038)

## Before this PR

The Ionide test explorer run on dotnet test and TRX files. 
Test discovery and test execution are implemented directly in Ionide.

## After this PR

Test discovery and execution are moved to the language server, allowing them to be used across clients.

Test functionality is also built on VSTestConsoleWrapper instead of `dotnet test`.
This comes with several advantages
- we now support Microsoft.Testing.Platform VSTestBridge projects
- We have access to test code locations provided by the test libraries themselves
- We don't have to run tests to discover them, which makes test discovery much faster
- We don't need to create temporary artifacts (i.e. TRX files)

We will not yet support Microsoft Testing Platform projects that don't use VSTestBridge.

## Outstanding issues

- I tried to support .runsettings ([ionide 2093](https://github.com/ionide/ionide-vscode-fsharp/discussions/2093) and [2040](https://github.com/ionide/ionide-vscode-fsharp/discussions/2040)), but there are [outstanding questions](https://github.com/ionide/FsAutoComplete/pull/1383#issuecomment-3258623066) on how to support it via VSTestConsoleWrapper.

- Microsoft Testing Platform VSTestBridge projects also report their test locations as the end of a test's code block.
There's not really a way for us to fix this.

- Additional work is needed to support pure Microsoft Testing Platform projects.




----
# Old draft PR text  

Short Version:  I'm unsure how to properly debug tests using Microsoft.TestPlatform.TranslationLayer. All the interactions with vstest are in [this file](https://github.com/ionide/FsAutoComplete/blob/41464460081e153d7ba43fc533f26e3459065cbd/src/FsAutoComplete.Core/TestServer.fs).

## Existing Behavior

Ionide is the F# tooling for VSCode. Ionide's test explorer currently works off of `dotnet test` with the TRX logger.
This approach will no longer work with the transition to Microsoft.Testing.Platform.

## Proposed Change

This PR moves Ionide's test platform integration to the language server and replaces `dotnet test` and trx files with [Microsoft.TestPlatform.TranlationLayer](https://www.nuget.org/packages/Microsoft.TestPlatform.TranslationLayer).

This should make Ionide more consistent with test explorers in other IDEs and hopefully sets us up for future work on dual vstest/MTP support.

So far I've got test discovering, running, and incremental updates streaming to the client.

## Current Challenges

**Debugging tests**
I'm stuck on running tests with the debugger attached.

I got a demonstration working by [parsing the processId from the log output](https://github.com/ionide/FsAutoComplete/blob/41464460081e153d7ba43fc533f26e3459065cbd/src/FsAutoComplete.Core/TestServer.fs#L68).
But, it appears that doesn't work with vstest versions 17.10.0 and earlier.

I also looked at `ITestRunEventsHandler.LaunchProcessWithDebuggerAttached` and `ITestHostLauncher`/`RunTestsWithCustomTestHost`, but I was not able to demonstrate any of the custom launch methods being called. I may be missing something hiding in plain sight.

**NUnit doesn't respect filters**

I'm using [this project](https://github.com/farlee2121/Ionide-Test-Explorer-RegressionTest) as a sample environment for testing both the old and new Ionide test explorer.

Filtering to a specific NUnit test using `dotnet test --filter "expression-here"` only logs results for the expected test.
However, TranslationLayer returns TestResults for all of the NUnit tests given the same filter expression.
TranslationLayer respects the filter expression for every other test library.

farlee2121 committed:
Set up the stub of a test explorer test project

farlee2121 committed:
Add the test explorer tests to the sln

farlee2121 committed:
Set up a trival test case to scaffold for VSTestAdapter work

farlee2121 committed:
Update Expecto for performance improvements

farlee2121 committed:
Demonstrate the most basic test discovery with vstest

farlee2121 committed:
Demonstrate connection between ionide and language server

farlee2121 committed:
Implement and mvp test discovery endpoint on the Language Server

farlee2121 committed:
Push test discovery down to AdaptiveState

This is preparation for adding notifications for incremental discovery reporting

farlee2121 committed:
Publish incremental test discovery updates

farlee2121 committed:
Demonstration a test run using VSTest

farlee2121 committed:
Rename the namespace for test features.

VSTestAdapter is too specific.
We already know we'll need to expand to accomodate Microsoft.Testing.Platform

farlee2121 committed:
Add an lsp endpoint for running tests

farlee2121 committed:
Fix bug where the TestExplorer tests don't build the required sample projects

farlee2121 committed:
Demonstrate basic streaming of test run results from langauge server

farlee2121 committed:
Add test filtering to the VSTest wrapper

farlee2121 committed:
Add testCaseFilter support to test Runs

Still has a bug where NUnit doesn't respect the filter, but works for all other frameworks

farlee2121 committed:
Prove that I can scrape test host process Ids

Also required figuring out graceful cancellation behavior for test runs

farlee2121 committed:
Convert VsTestWrapper methods to async-only for graceful cancellation

farlee2121 committed:
Pass a server-level test for the attaching debugger to test run.

farlee2121 committed:
Format files

farlee2121 committed:
Delete some diagnostic code

Youssef1313 commented on src/FsAutoComplete.Core/TestServer.fs:
```diff
@@ -0,0 +1,223 @@
+                vstest.CancelTestRun()
+                printfn "Test Run Cancelled")
+
+            vstest.RunTests(sources, null, options, runHandler) 
```
@nohwnd is the best to help with VSTest stuff. But, I see there is a `RunTestsWithCustomTestHost` API on `VsTestConsoleWrapper`. From there, you have `LaunchTestHost` method with `TestProcessStartInfo` parameter, so that you can start the process yourself and then you know its id. That should be more reliable than parsing process id from the logs, which you stated doesn't work with older versions of VSTest.

nohwnd replied:
https://github.com/microsoft/vstest/blob/main/playground/TestPlatform.Playground/Program.cs#L116 this is how you'd do it. The playground project in VSTest is also the best way to experiment with this kind of stuff. If you build, you will have attachVS tool available that we build in the same repo, this tool will automatically attach your VisualStudio to the processes that you choose. E.g. setting `VSTEST_RUNNER_DEBUG_ATTACHVS` to 1, in this file  https://github.com/microsoft/vstest/blob/main/playground/TestPlatform.Playground/Environment.cs#L14 will attach vs to vstest.console process automatically so you can debug both the client and the server.

In vstest.console you are interested in DesignMode client, that is where the messages from vstestconsolewrapper are processed. 

https://github.com/microsoft/vstest/blob/main/src/Microsoft.TestPlatform.Client/DesignMode/DesignModeClient.cs#L162

For debugging there are 2 possible flows, launch with debugger attached (this would happen before testhost starts), and attach debugger (this is driven from the testhost after it starts). VS is using AttachDebugger, which is more modern. But it depends on you what you use.

<https://github.com/microsoft/vstest/blob/main/docs/Overview.md> here is documentation of the protocol, if you search there for "debugger" you will hopefully find the info you need, and the description of the flow, and the versions of interfaces that support it.

But honestly this is not something that people do every day, so that protocol description might be lacking all the info you need to successfully implement this. Feel free to ask more questions. :)

nohwnd replied:
https://github.com/microsoft/vstest/blob/main/playground/TestPlatform.Playground/Program.cs#L252 there you probably want to implement ITestHostLauncher3, that also gives you info about tfm, and the dlls you will debug. 

This is interesting for multi target framework runs, where the debugger might differ between .net framework and .net.

https://github.com/microsoft/vstest/blob/main/src/Microsoft.TestPlatform.ObjectModel/Client/Interfaces/ITestHostLauncher3.cs

https://github.com/microsoft/vstest/blob/main/src/Microsoft.TestPlatform.ObjectModel/Client/Interfaces/AttachDebuggerInfo.cs#L8

farlee2121 replied:
Thanks for the help!

`ITestHostLauncher3` was the solution to my debugging challenges.
`RunTestsWithCustomTestHost` only has overloads taking `ITestHostLauncher`, so I overlooked ITestHostLauncher2 and ITestHostLauncher3. Poking around the TestPlatform code, it looks like there's some casting going on.

farlee2121 replied:
Also, the TestPlatform playground has been indispensable as I've tried to learn how to work with VSTest.
I'm grateful to the team for open sourcing it!

I extracted a [standalone copy of the TestPlatform Playground](https://github.com/farlee2121/TestPlatform.Playground). That's where I run most of my experiments

nohwnd replied:
Sweet! The playground  and attachvs is my work and I find it super useful almost every day as well, glad to see I am not the only one  👍

Yeah that api is not great, we do awful lot to avoid breaking public api and break users. And also don't want to add more and more copies of the public api with all the combinations of possible new interfaces. So keeping the old interface in the api, and then adapting it internally to the newest version, if needed, is the way that multiple things are implemented. But it leads to making it difficult to discover.

Youssef1313 commented:
For NUnit, this is more likely to be an NUnit bug. I see strange filtering logic here.

https://github.com/nunit/nunit3-vs-adapter/blob/80a67d734bb1e9144b16c67cd61e79d1d14d9349/src/NUnitTestAdapter/NUnit3TestExecutor.cs#L134-L142

https://github.com/nunit/nunit3-vs-adapter/blob/80a67d734bb1e9144b16c67cd61e79d1d14d9349/src/NUnitTestAdapter/NUnit3TestExecutor.cs#L172-L182

When running with VsTestConsoleWrapper, DesignMode will be true, and that's likely why this is different from `dotnet test` and also why it's NUnit-specific. In MSTest for example, I don't think we treat "DesignMode" differently anywhere.

I think what might be worth trying is to embed the filter in runsettings, instead of TestPlatformOptions. Meanwhile, I think it's worth reporting this bug to NUnit.

farlee2121 committed:
fix incorrect message type on the runTests lsp message

farlee2121 committed:
Pass unit tests using RunTestsWithCustomTestHost

RunTestsWithCustomTestHost is the official way to attach a debugger with VSTest

farlee2121 commented:
@Youssef1313 I agree the filter issue is probably something to report to NUnit. 

I tried your idea to set the filter in runsettings, but it didn't seem to change the behavior. 
I'll need to poke around the NUnit code you referenced.

Youssef1313 commented:
> I tried your idea to set the filter in runsettings, but it didn't seem to change the behavior.

There are two things you can try with runsettings. First is forcing DesignMode to be false. Second is passing the filter via TestCaseFilter. If passing via TestCaseFilter doesn't work too, then that's yet another NUnit bug.

Youssef1313 commented:
MSTest handles that here:

https://github.com/microsoft/testfx/blob/a8bd57bd6f5eefbde2caf8ad9857436083da6cbb/src/Adapter/MSTestAdapter.PlatformServices/TestMethodFilter.cs#L124-L126

xUnit.net handles that here:

https://github.com/xunit/visualstudio.xunit/blob/e06d9e3caee1a08dec30f8164a8ae280fcdf5667/src/xunit.runner.visualstudio/Utility/TestCaseFilter.cs#L131

@nohwnd btw, why do frameworks need to do reflection here? Is there no public API that exposes the needed info to adapters?

spencer-wcri commented:
Setting `<DesignMode>False</DesignMode>` does fix the NUnit filtering behavior.

This tag [doesn't appear in the official documentation](https://learn.microsoft.com/en-us/visualstudio/test/configure-unit-tests-by-using-a-dot-runsettings-file?view=vs-2022#runconfiguration-element), do you think it'd be risky to use?

Youssef1313 commented:
@nohwnd is the best to answer here.

farlee2121 committed:
Debug test runs

Broke out the debug attach message because it needs to be a blocking call.
VSTest will continue executing the tests as soon as AttachDebuggerToProcess returns

I also fixed the LSP level test explorer tests, which were failing because the sample test projects needed to be built at part of the unit test, since the build artifacts are being cleared before every test run

farlee2121 committed:
Test that client environment variables are available in a test run

farlee2121 committed:
Decouple SampleTestProjects for Lsp and TestExplorer test projects

FsAutoComplete.Tests.TestExplorer expects the sample projects to be built once before the tests are run,
but FsAutoComplete.Tests.Lsp expects the sample projects to be build for every test.
Splitting up the sample projects reduces chances that they could break each other when the run in parallel or in random order

farlee2121 committed:
Forward test logs

farlee2121 committed:
Allow client to limit which test projects are run

This was mainly motivated by test debugging.

Before this feature, the debugger would launch for every test project even if you only wanted to debug one test.
With this change, it'll only launch the one project.

This also partially mitigates the bug where NUnit doesn't respect filters, now those test won't show unless some tests from an nunit project are selected

farlee2121 committed:
Workaround for NUnit not respecting test filters

NUnit doesn't respect test filters when VSTest is in DesignMode, which it is by default with VsTestConsoleWrapper

https://github.com/ionide/FsAutoComplete/pull/1383#issuecomment-3245590606

farlee2121 commented:
I'm looking to add .runsettings support to the Ionide test explorer.
I see that the playground [has an example of runsettings](https://github.com/microsoft/vstest/blob/f7325aaca89a8391520519f97cd749a8be281544/playground/TestPlatform.Playground/Program.cs#L41).

It appears `VsTestConsoleWrapper` only accepts runsettings xml (it complains about invalid xml if I try pass a path), but some runsettings elements are supposed to be paths relative to the runsettings file (i.e. [ResultsDirectory](https://learn.microsoft.com/en-us/visualstudio/test/configure-unit-tests-by-using-a-dot-runsettings-file?view=vs-2022#runconfiguration-element)). This behavior can't be satisfied if vstest doesn't know where the runsettings came from, and I don't see a way to specify where the runsettings came from or another way of passing run settings.

Is there another way to support runsettings that accounts for the runsettings file location?

farlee2121 committed:
Forward test discovery logs to the client for improved error diagnostics

farlee2121 commented:
@nohwnd Do you have some ideas on the [.runsettings challenges with VSTestConsoleWrapper](https://github.com/ionide/FsAutoComplete/pull/1383#issuecomment-3258623066)

farlee2121 committed:
Show error notification in client for expectable test discovery issues (i.e projects not restored or built)

farlee2121 committed:
Separate discovery vs run lsp-level test explorer tests

farlee2121 committed:
Split the VSTest wrapper code from TestServer contracts for clarity

farlee2121 committed:
Clarify dictionary name (projectLookup -> projectsByBinaryPath)

farlee2121 committed:
Satisfy fantomas formatting check

farlee2121 committed:
Remove print statements per copilot review

TheAngryByrd commented on src/FsAutoComplete/LspServers/AdaptiveFSharpLspServer.fs:
```diff
@@ -3056,6 +3056,24 @@ type AdaptiveFSharpLspServer
 
+    override this.TestDiscoverTests() : Async<LspResult<PlainNotification option>> =
+      asyncResult {
+        let! testDTOs =
```
Do add OTel/Logging like other methods in this class.

TheAngryByrd commented on src/FsAutoComplete/LspServers/AdaptiveServerState.fs:
```diff
@@ -156,6 +156,29 @@ type FindFirstProject() =
+      |> List.ofSeq
+      |> List.map string
+      |> workspaceLoader.LoadProjects
+      |> Ok
```
```suggestion

```

Instead of loading projects again, you'll want to reuse the currently loaded projects as this will help with any project related changes like adding files, adding packages etc.

https://github.com/ionide/FsAutoComplete/blob/20167d6d65aedc7f641cb27dd571c2d2c0830f50/src/FsAutoComplete/LspServers/AdaptiveServerState.fs#L944

You can do something like:

```fsharp
let! projects = projectOptions |> AsyncAval.forceAsync
let! testProjects = projectOptions.ToValueSeq() |> TestProjectHelpers.tryGetTestProjects
```

In DiscoverTests/RunTests

TheAngryByrd commented on src/FsAutoComplete.Core/VSTestWrapper.fs:
```diff
@@ -0,0 +1,191 @@
+          this.DiscoveredTests.AddRange(lastChunk)
+          notifyDiscoveryProgress (lastChunk |> List.ofSeq |> Progress)
+
+      member this.HandleLogMessage(level: TestMessageLevel, message: string) : unit =
```
Is it worth logging the message on the server side as well?

TheAngryByrd reviewed (changes requested):
Couple of minor changes. This is looking great!

farlee2121 committed:
Add OpenTelemetry to the Test Explorer langauge server endpoints

farlee2121 committed:
Avoid duplicate project load when discovery test projects

farlee2121 commented:
I looked into the test failures.

It appears that a random group of tests in AdjustConstantTests semi-regularly fails.
It doesn't appear to consistently fail a particular test group or set of groups. 

The error is always thrown from `waitForLatestDiagnostics`.
This method comes with a big note discussing the tradeoff between test time and chance of timeout.

I assume this timeout is why random tests in the group are intermittently failing.
I tried to play with the relevant timeout (`AdjustConstantTests > check > Server.createUntitledDocument > Document.openWith`), but I didn't see a significant improvement in test stability with small increases in the timeout

TheAngryByrd commented:
> The error is always thrown from waitForLatestDiagnostics.
> This method comes with a big note discussing the tradeoff between test time and chance of timeout.

Yeah, we technically could use [PullDiagnostics](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_pullDiagnostics) but I'd fear we would possibly regress on the push diagnostics part. This has definitely been a thing that's been noisy for years tho.

TheAngryByrd committed:
Fix net9 build
